File: ImplementHashset.java
// Time Complexity : O(1) on average but in the worst case it could be O(n) for add, remove contains;
// Space Complexity : O(n) depending on the number of entries in the hashset
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Customized the linked list + array approach from yesterday's class


// Your code here along with comments explaining your approach

class MyHashSet {
    
    private class Node{
        int value;
        Node next;
        
        Node(int value){
            this.value = value;
        }
    }
    
    Node[] nodes;
    
    private int findIndex(int value){
        return Integer.hashCode(value) % nodes.length;
    }
    
    private Node findElement(int index, int value){
        if(nodes[index] == null){
            return nodes[index] = new Node(-1);
        }
        else{
            Node prev = nodes[index];
            while(prev.next != null && prev.next.value != value){
                prev = prev.next;
            }
            return prev;
        }
    }

    /** Initialize your data structure here. */
    public MyHashSet() {
        nodes = new Node[1000001];
    }
    
    public void add(int key) {
        int index = findIndex(key);
        Node prev = findElement(index, key);
        
        if(prev.next != null){
            prev.next.value = key;
        }
        else{
            prev.next = new Node(key);
        }        
    }
    
    public void remove(int key) {
        int index = findIndex(key);
        Node prev = findElement(index, key);
        
        if(prev.next != null){
            prev.next = prev.next.next;
        } 
    }
    
    /** Returns true if this set contains the specified element */
    public boolean contains(int key) {
        int index = findIndex(key);
        Node prev = findElement(index, key);
        
        if(prev.next != null && prev.next.value == key){
            return true;
        } 
        else{
            return false;
        }
    }
}

File: ImplementHashSet.py
# Time Complexity - O(N)
#Space Complexity - O(N)

# Similar to HashMap implementation
# ListNode to store key and value
class ListNode:
    def __init__(self,key):
        self.key = key
        self.next = None
        
class MyHashSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.size = 10000
        self.nodes = [None]*self.size
        
    # Hashing funtion
    def getHash(self,key):
        return key%self.size
    
    def findElement(self,key,head):
        prev=None
        cur = head
        
        while(cur != None and cur.key!=key):
            prev=cur
            cur=cur.next
            
        return prev
        

    def add(self, key: int) -> None:
        
        hash_value = self.getHash(key)
        if self.nodes[hash_value]==None:
            self.nodes[hash_value]=ListNode(-1) # Creating a dummy node at the beginning
        prev = self.findElement(key,self.nodes[hash_value])
        
        # If key does not exist, then create a new ListNode with given key and value as next of prev node
        if prev.next==None:
            prev.next = ListNode(key)
        

    def remove(self, key: int) -> None:
        
        hash_value = self.getHash(key)
        if self.nodes[hash_value] == None:
            return
        prev = self.findElement(key,self.nodes[hash_value])
        
        if prev.next==None:
            return
        
        prev.next=prev.next.next
        
        

    def contains(self, key: int) -> bool:
        """
        Returns true if this set contains the specified element
        """
        hash_value = self.getHash(key)
        
        # base case 
        if self.nodes[hash_value] == None:
            return False
        prev = self.findElement(key,self.nodes[hash_value])
        
        if prev.next == None:
            return False
        
        return True
        
        


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)